docs/source/development/controllers.rst

```diff
@@ -0,0 +1,221 @@
+inputs, we can test a controller by mocking the hardware dependencies and
+asserting expected outputs are sent to particular hardware.
+
+We are currently working on a testing framework for eVolver controllers (please
+see https://github.com/ssec-jhu/evolver-ng/issues/156), but in the meantime, an
+example for mocking hardware within an eVolver environment can be seen in the
+`test_chemostat.py` file in the eVolver repository.
+
+
+Logging in the controller
+-------------------------
+
+All components in the eVolver framework contain in internal logger which is an
+instance of a python standard library `logging.Logger`. This logger can be used
+to emit messages from within a controller which, unless otherwise configured,
+will get routed through the eVolver logging mechanism.
+
+.. note::
+
+    At present, the logger is configured only with basic handling, and will
+    print to stdout. In future releases more advanced logging and event handling
+    is planned, along with retreival of logs from the API.
+
+
+Example::
+
+    class MyController(Controller):
+        def control(self):
+            self.logger.info('Starting control loop')
+            for vial in self.vials:
+                self.logger.debug(f'Processing vial {vial}')
+                ...
+
+
+History
+-------
+
+Some experiments may require access to historical data in order to make control
+decisions. While it is always possible to store a buffer of historic data for a
+given sensor in memory within the controller, this may have unintended
+consequences in the case of reboot or even a reconfiguration of the eVolver
+(which reallocates all objects): the buffer would be lost.
+
+eVolver provides a history server which is backed by persistent storage designed
+to be queried via the API or within a controller. This means controllers code
+can remain simple and focus on core logic, as opposed to maintaining file-based
+historic data or error-prone in-memory buffers.
+
+To access the history server, the controller can use the `self.evolver.history`
+property which is automatically available to all controllers when used within
+the application (see :ref:`evolver-hookup` below).
+
+For example::
+
+        class MyController(Controller):
+            def control(self):
+                # Get the temperature history for all vials for past hour
+                temp_history = self.evolver.history.get('temp', t_start=time.time() - 3600)
+                for vial in self.vials:
+                    mean_temp = np.mean([i.data[0]['value'] for i in  history.get(name='test').data['test']])
+
+.. note::
+
+    The response from the history server
+    :py:class:`evolver.history.HistoryResult` is amenable to transport over the
+    network, but presently has overhead in working with results in the
+    interpreter (as can be noted in the above example). Functionality like
+    `to_dataframe` or `to_numpy` can be provided in future versions.
```

Comment:
A quick additional note that history isn't namespaced by controller and thus it's the dev's responsibility to pre or post filter history data for the vials specified by and controlled by the given controller.